### ISSUE TITLE
Shortcuts not working if default WSL shell is fish

### ISSUE THREAD

**Issue by sanzoghenzo** [Opticos/GWSL-Source#14]
Hi there, first of all thank you so much for this tool!

After some trial and error I discovered that if I set fish as my default shell in WSL2, then the shortcuts don't work.
I can see some processes spawning and then exiting, no log is produced.

The same happens if I use cmd to run the shortcut command.

switching back to bash "solved" the problem.

I tried to run the various commands and script directly in my fish shell, here's what I found:

The first problem is in the get_ip function:

```shell
$ /bin/sh -c "echo $(cat /etc/resolv.conf | grep nameserver | awk '{print $2; exit;}')"
fish: $(...) is not supported. In fish, please use '(cat)'.
```

I set my DISPLAY in both `.profile` and `.config/fish/config.fish`, so maybe the ip discovery and DISPLAY setting can be omitted [here](https://github.com/Opticos/GWSL-Source/blob/016872c5f14c240bbe5b228bbb92d271feac842a/manager.py#L1921-L1922) (but I suppose it will break many things).

also setting the envvars doesn't work; in my simple case (no GTK theme set, dark windows theme and everything default):

```shell
$ GTK_THEME=Adwaita:dark /opt/pycharm-2020.2.3/bin/pycharm.sh                  (base)
fish: Unsupported use of '='. To run '/opt/pycharm-2020.2.3/bin/pycharm.sh' with a modified environment, please use 'env GTK_THEME=Adwaita:dark /opt/pycharm-2020.2.3/bin/pycharm.sh…'
```

In short, it would be wonderful if you can also add support to fish!

BTW, reading your code I saw many things that could be improved; are you open to code reviews/PRs? If so, can you please describe your build environment?
Of course, I could also try to fix this issue myself once I know which tools you use

**Comment by Pololot64** [comment]
Hello! Thanks for all the work you have done so far! Yes, unfortunately when I started this project I had no idea that people used things other than bash. I hope to have them supported and I am open to contributions. I am thinking that a sort of addon system would be good. Pull requests are good and code review is good too. I am aware much of the code is messy :). Part of it is due to the fact that I scrambled to finish it and fix some bugs before summer ended. I have not been able to change much during the school year.

I do not have a list of dependencies written down yet but that can all be installed via pip. They can be found by trial and error I guess but I will make this more clear.

The structure of the program is relatively simple. The file manager.py handles the dashboard gui and shortcut functionality. The file main.py is the service that runs in the notification tray and handles the main vcxsrv instance.

I build it with pyinstaller at the moment. I first bundle main as a onefolder in noconsole mode and then bundle manager. I copy the generated binaries into one folder and ignore file conflicts. They share dependencies. Then copy all the assets other than the original python files into that folder. Running Manager.exe should start the service and pull up the dashboard. Packaging the program for the store is a whole other story and requires other steps but it does not directly affect the rest of the program.

I do not have my development computer on me at the moment so I am writing this from memory but I hope to soon have it documented.

**Comment by sanzoghenzo** [comment]
Hi @Pololot64,

I started to analyze and reorganize your code, but I think it's trying to do too many things at once.

So I took another approach, and I hope you don't mind: I extracted and organized the wsl management part into a separate library called [wsl-tools](https://github.com/sanzoghenzo/wsl-tools).
I think it could be useful also to other projects.

You can install it via `pip install wsl-tools`. 

I still have to publish the api reference on github pages, but you can find the information in the docstrings and in the tests.

here's a quick example:

```py
from wsl_tools import wsl_tools

manager = wsl_tools.WSLManager()
# manager.names holds all the non-blacklisted distros names
distro_name = choose_machine(manager.names)
# access the distro using the name as key
distro = manager[distro_name]
# access the installed GUI apps names
app = choose_app(distro.gui_apps.keys())
# run the app loading .profile first
distro.run_command(app.cmd, load_profile=True)
```

I found out that you can ignore the default shell and run `sh -lc {command}` to run sh (=bash on debian or ash on alpine)  and load the .profile, so I kept the various settings there.
I also got rid of the helper script and tried to do as much as I could in the WSLDistro class: the various envvar do get from/set to the .profile are properties of that class (gtk_scale, qt_scale, theme, etc.).

I didn't try to use wsl-tools in gwsl yet  (and I still need to test gui apps with it!), but if you like this approach I could integrate it along with other things (a settings manager, path manager, cli argument parsing, etc).

**Comment by Pololot64** [comment]
Wow! This is amazing! I am going to test it as soon as I am able! I am wondering though, how did you get xdg.DesktopEntry to work on Windows? The reason I wrote the whole messy script was because I could not get it to run.

**Comment by sanzoghenzo** [comment]
The trick is done in [`_unc_path_from_cmd` method](https://github.com/sanzoghenzo/wsl-tools/blob/b276355f0cc12b65e9cb8e1709d91d9ced4ed828/src/wsl_tools/__init__.py#L163): with it you can convert the wsl POSIX path to the `\\wsl$\distro-name` path, so you can access and edit the files directly in python.

The method also follows the symlinks, because windows can't do it;
In fact, the .desktop files retrieval has a limitation: since I only use the method to get the directory and then glob all the .desktop files, any entry that is a symlink cannot be read, and it's ignored at the moment. (I should file an issue to remember to fix this!)

**Comment by Pololot64** [comment]
I see! I never realized that there was a pyxdg 🤦‍♂️. By the way, alot of this could be integrated but it would take some work. Would you want to create a quicker contact system? Maybe the Discord server?

**Comment by sanzoghenzo** [comment]
ok, I'll meet you there!
